File: output/gemini-3.1-pro-preview/zero/cpp/Burocracia.cpp

#include <iostream>
#include <vector>

using namespace std;

const int MAXN = 100005;
const int INF = 1e9;

int N, Q;
int p[MAXN];
vector<int> adj[MAXN];

int up[MAXN][20];
int depth[MAXN];
int in[MAXN], out[MAXN];
int timer = 0;

int tree[4 * MAXN];

void dfs(int u, int p_node, int d) {
    in[u] = ++timer;
    depth[u] = d;
    up[u][0] = p_node;
    for (int i = 1; i < 20; i++) {
        up[u][i] = up[up[u][i-1]][i-1];
    }
    for (int v : adj[u]) {
        if (v != p_node) {
            dfs(v, u, d + 1);
        }
    }
    out[u] = timer;
}

void update(int node, int l, int r, int ql, int qr, int u) {
    if (ql > r || qr < l) return;
    if (ql <= l && r <= qr) {
        if (depth[u] < depth[tree[node]]) {
            tree[node] = u;
        }
        return;
    }
    int mid = (l + r) / 2;
    update(2 * node, l, mid, ql, qr, u);
    update(2 * node + 1, mid + 1, r, ql, qr, u);
}

int query(int node, int l, int r, int pos) {
    int res = tree[node];
    if (l == r) return res;
    int mid = (l + r) / 2;
    int child_res;
    if (pos <= mid) {
        child_res = query(2 * node, l, mid, pos);
    } else {
        child_res = query(2 * node + 1, mid + 1, r, pos);
    }
    if (depth[child_res] < depth[res]) {
        res = child_res;
    }
    return res;
}

int get_kth_ancestor(int u, int k) {
    for (int j = 0; j < 20; j++) {
        if (k & (1 << j)) {
            u = up[u][j];
        }
    }
    return u;
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    if (!(cin >> N)) return 0;

    for (int i = 2; i <= N; i++) {
        cin >> p[i];
        adj[p[i]].push_back(i);
    }

    depth[0] = INF;
    dfs(1, 1, 0);

    cin >> Q;
    while (Q--) {
        int type;
        cin >> type;
        if (type == 1) {
            int v, k;
            cin >> v >> k;
            int h = query(1, 1, N, in[v]);
            if (h == 0) {
                cout << get_kth_ancestor(v, k) << "\n";
            } else {
                if (k == 1) {
                    cout << h << "\n";
                } else {
                    cout << get_kth_ancestor(h, k - 1) << "\n";
                }
            }
        } else {
            int v;
            cin >> v;
            if (in[v] + 1 <= out[v]) {
                update(1, 1, N, in[v] + 1, out[v], v);
            }
        }
    }

    return 0;
}

File: output/cbsoft_sbes_2026/google_gemini_3_1_pro_preview/zero/python_noimg/Burocracia.py
import sys

def solve():
    input = sys.stdin.read
    data = input().split()
    if not data:
        return
    
    N = int(data[0])
    
    p = [0] * (N + 1)
    for i in range(2, N + 1):
        p[i] = int(data[i - 1])
        
    Q_idx = N
    Q = int(data[Q_idx])
    
    LOG = 18
    up = [0] * ((N + 1) * LOG)
    depth = [0] * (N + 1)
    
    adj = [[] for _ in range(N + 1)]
    for i in range(2, N + 1):
        up[i * LOG + 0] = p[i]
        adj[p[i]].append(i)
        
    for i in range(2, N + 1):
        depth[i] = depth[p[i]] + 1
        for j in range(1, LOG):
            prev = up[i * LOG + j - 1]
            up[i * LOG + j] = up[prev * LOG + j - 1]
            
    in_time = [0] * (N + 1)
    out_time = [0] * (N + 1)
    
    edge_idx = [0] * (N + 1)
    stack = [1]
    timer = 0
    in_time[1] = timer
    timer += 1
    
    while stack:
        u = stack[-1]
        if edge_idx[u] < len(adj[u]):
            v = adj[u][edge_idx[u]]
            edge_idx[u] += 1
            in_time[v] = timer
            timer += 1
            stack.append(v)
        else:
            out_time[u] = timer - 1
            stack.pop()
            
    INF = 10**9
    M = N + 1
    tree = [INF] * (2 * M)
    
    def update(l, r, val):
        l += M
        r += M + 1
        while l < r:
            if l % 2 == 1:
                if val < tree[l]: tree[l] = val
                l += 1
            if r % 2 == 1:
                r -= 1
                if val < tree[r]: tree[r] = val
            l //= 2
            r //= 2

    def query(pos):
        pos += M
        res = INF
        while pos > 0:
            if tree[pos] < res: res = tree[pos]
            pos //= 2
        return res

    def get_kth_ancestor(u, k):
        for j in range(LOG):
            if (k >> j) & 1:
                u = up[u * LOG + j]
        return u

    idx = Q_idx + 1
    out = []
    for _ in range(Q):
        type = int(data[idx])
        if type == 1:
            v = int(data[idx+1])
            k = int(data[idx+2])
            idx += 3
            
            pv = p[v]
            d = query(in_time[pv])
            if d == INF:
                ans = get_kth_ancestor(v, k)
            else:
                diff = depth[pv] - d
                y = pv
                if diff > 0:
                    y = get_kth_ancestor(pv, diff)
                if k == 1:
                    ans = y
                else:
                    ans = get_kth_ancestor(y, k - 1)
            out.append(str(ans))
        else:
            v = int(data[idx+1])
            idx += 2
            update(in_time[v], out_time[v], depth[v])
            
    sys.stdout.write('\n'.join(out) + '\n')

if __name__ == '__main__':
    solve()